### ISSUE TITLE
java is broken

### ISSUE THREAD

**Issue by spuder** [Appdynamics/appdynamics-cookbooks#64]
Getting the following when trying to deploy a java application

```
================================================================================
           Error executing action `put` on resource 'ark[javaagent]'
           ================================================================================

           Addressable::URI::InvalidURIError
           ---------------------------------
           Invalid port number: "\\Users\\vagrant\\AppData\\Local\\Temp\\kitchen\\cache"

           Resource Declaration:
           ---------------------
           # In C:/Users/vagrant/AppData/Local/Temp/kitchen/cache/cookbooks/appdynamics/recipes/java_agent.rb

            23: ark 'javaagent' do
            24:   url "file://#{node['appdynamics']['java_agent']['zip']}"
            25:   path agent['install_dir']
            26:   owner agent['owner']
            27:   action :put
            28: end
            29:

           Compiled Resource:
           ------------------
           # Declared in C:/Users/vagrant/AppData/Local/Temp/kitchen/cache/cookbooks/appdynamics/recipes/java_agent.rb:23:in `from_file'

           ark("javaagent") do
             action [:put]
             retries 0
             retry_delay 2
             default_guard_interpreter :default
             declared_type :ark
             cookbook_name "appdynamics"
             recipe_name "java_agent"
             url "file://C:\\Users\\vagrant\\AppData\\Local\\Temp\\kitchen\\cache/AppDynamicsJavaAgent.zip"
             path "c:\\program files\\appdynamics\\javaagent/javaagent"
             owner "root"
             extension "zip"
             release_file "C:\\Users\\vagrant\\AppData\\Local\\Temp\\kitchen\\cache/javaagent.zip"
           end

           Platform:
           ---------
           x64-mingw32

```

kitchen.yml

```yml
  - name: java
    run_list:
    - recipe[appdynamics::default]
    - recipe[appdynamics::java_agent]
    attributes:
      appdynamics:
        java_agent:
          version: 4.1.8.9
          checksum: bcfb91da7ee74358b902e260825d427b5901c961a2179be751430e95c46225ec
          install_dir: c:\program files\appdynamics\javaagent
        controller:
          host: host.example.com
          port: 443
          ssl: tru
          user: fooobar
          accesskey: xxxxxx
        version: 4.1.8.9
```

**Comment by dkoepke** [comment]
Please try setting `#{node['appdynamics']['java_agent']['zip']}` to be `/#{Chef::Config[:file_cache_path]}/AppDynamicsJavaAgent.zip` (note the leading slash).

**Comment by spuder** [comment]
I assume you are referring to this line? 

https://github.com/Appdynamics/appdynamics-cookbooks/blob/master/recipes/java_agent.rb#L24

**Comment by dkoepke** [comment]
You can set this attribute on a node

**Comment by spuder** [comment]
I've added to my kitchen.yml

```
    attributes:
      appdynamics:
        packages_site: https://packages.netdocuments.com:443/artifactory/nd-appdynamics
        java_agent:
          version: 4.1.8.9
          checksum: bcfb91da7ee74358b902e260825d427b5901c961a2179be751430e95c46225ec
          install_dir: c:\appdynamics
          zip: /#{Chef::Config[:file_cache_path]}/AppDynamicsJavaAgent.zip
```

getting

```
           ================================================================================
           Error executing action `create` on resource 'remote_file[/#{Chef::Config[:file_cache_path]}/AppDynamicsJavaAgent.zip]'
           ================================================================================

           Chef::Exceptions::EnclosingDirectoryDoesNotExist
           ------------------------------------------------
           Parent directory /#{Chef::Config[:file_cache_path]} does not exist.

           Resource Declaration:
           ---------------------
           # In C:/Users/vagrant/AppData/Local/Temp/kitchen/cache/cookbooks/appdynamics/recipes/java_agent.rb

            16: remote_file node['appdynamics']['java_agent']['zip'] do
            17:   source package_source
            18:   checksum agent['checksum']
            19:   backup false
            20:   mode '0444'
            21: end
            22:

           Compiled Resource:
           ------------------
           # Declared in C:/Users/vagrant/AppData/Local/Temp/kitchen/cache/cookbooks/appdynamics/recipes/java_agent.rb:16:in `from_file'
```

I tried changing my berksfile to point to a local copy and modified line 24 directly

```
ark 'javaagent' do
  url "file://#{Chef::Config[:file_cache_path]}/AppDynamicsJavaAgent.zip"
  path agent['install_dir']
  owner agent['owner']
  action :put
end
```

Getting similar error

```
         * ark[javaagent] action put[2016-12-21T21:38:17+00:00] WARN: file://C:\Users\vagrant\AppData\Local\Temp\kitchen\cache/AppDynamicsJavaAgent.zip was an invalid URI. Trying to escape invalid characters


           ================================================================================
           Error executing action `put` on resource 'ark[javaagent]'
           ================================================================================

           Addressable::URI::InvalidURIError
           ---------------------------------
           Invalid port number: "\\Users\\vagrant\\AppData\\Local\\Temp\\kitchen\\cache"

           Resource Declaration:
           ---------------------
           # In C:/Users/vagrant/AppData/Local/Temp/kitchen/cache/cookbooks/appdynamics/recipes/java_agent.rb

            23: ark 'javaagent' do
            24:   url "file://#{Chef::Config[:file_cache_path]}/AppDynamicsJavaAgent.zip"
            25:   path agent['install_dir']
            26:   owner agent['owner']
            27:   action :put
            28: end
            29:

           Compiled Resource:
           ------------------
           # Declared in C:/Users/vagrant/AppData/Local/Temp/kitchen/cache/cookbooks/appdynamics/recipes/java_agent.rb:23:in `from_file'

           ark("javaagent") do
             action [:put]
             retries 0
             retry_delay 2
             default_guard_interpreter :default
             declared_type :ark
             cookbook_name "appdynamics"
             recipe_name "java_agent"
             url "file://C:\\Users\\vagrant\\AppData\\Local\\Temp\\kitchen\\cache/AppDynamicsJavaAgent.zip"
             path "c:\\appdynamics/javaagent"
             owner "root"
             extension "zip"
             release_file "C:\\Users\\vagrant\\AppData\\Local\\Temp\\kitchen\\cache/javaagent.zip"
           end

           Platform:
           ---------
           x64-mingw32

```

**Comment by dkoepke** [comment]
Ah, your `kitchen.yml` doesn't let you interpolate. If you change line 24 directly, it needs to have _three_ slashes:

```
url "file:///#{Chef::Config[:file_cache_path]}/AppDynamicsJavaAgent.zip"
```

**Comment by spuder** [comment]
Tried that too. Looks like It got a little further


```
         * remote_file[C:\Users\vagrant\AppData\Local\Temp\kitchen\cache/AppDynamicsJavaAgent.zip] action create[2016-12-21T21:55:39+00:00] WARN: Mode 444 includes bits for the owner, but owner is not specified
       [2016-12-21T21:55:39+00:00] WARN: Mode 444 includes bits for the group, but group is not specified
        (up to date)
         * ark[javaagent] action put[2016-12-21T21:55:39+00:00] WARN: file:///C:\Users\vagrant\AppData\Local\Temp\kitchen\cache/AppDynamicsJavaAgent.zip was an invalid URI. Trying to escape invalid characters

           * directory[c:\appdynamics/javaagent] action create (up to date)
           * remote_file[C:\Users\vagrant\AppData\Local\Temp\kitchen\cache/javaagent.zip] action create[2016-12-21T21:55:39+00:00] WARN: file:///C:\Users\vagrant\AppData\Local\Temp\kitchen\cache/AppDynamicsJavaAgent.zip was an invalid URI. Trying to escape invalid characters

             - create new file C:\Users\vagrant\AppData\Local\Temp\kitchen\cache/javaagent.zip
             - update content in file C:\Users\vagrant\AppData\Local\Temp\kitchen\cache/javaagent.zip from none to bcfb91
             (file sizes exceed 10000000 bytes, diff output suppressed)
           * execute[unpack C:\Users\vagrant\AppData\Local\Temp\kitchen\cache/javaagent.zip] action nothing (skipped due to action :nothing)
           * execute[set owner on c:\appdynamics/javaagent] action nothing (skipped due to action :nothing)
           * execute[unpack C:\Users\vagrant\AppData\Local\Temp\kitchen\cache/javaagent.zip] action run

             ================================================================================
             Error executing action `run` on resource 'execute[unpack C:\Users\vagrant\AppData\Local\Temp\kitchen\cache/javaagent.zip]'
             ================================================================================

             Mixlib::ShellOut::ShellCommandFailed
             ------------------------------------
             Expected process to exit with [0], but received '1'
             ---- Begin output of "C:\7-zip\7z.exe" e "C:\Users\vagrant\AppData\Local\Temp\kitchen\cache/javaagent.zip" -o"C:/Users/vagrant/AppData/Local/Temp/d20161221-2696-10pu6l" -uy && for /f %1 in ('dir /ad /b "C:\Users\vagrant\AppData\Local\Temp\d20161221-2696-10pu6l"') do xcopy "C:\Users\vagrant\AppData\Local\Temp\d20161221-2696-10pu6l\%1" "" /s /e ----
             STDOUT:
             STDERR: The system cannot find the path specified.
             ---- End output of "C:\7-zip\7z.exe" e "C:\Users\vagrant\AppData\Local\Temp\kitchen\cache/javaagent.zip" -o"C:/Users/vagrant/AppData/Local/Temp/d20161221-2696-10pu6l" -uy && for /f %1 in ('dir /ad /b "C:\Users\vagrant\AppData\Local\Temp\d20161221-2696-10pu6l"') do xcopy "C:\Users\vagrant\AppData\Local\Temp\d20161221-2696-10pu6l\%1" "" /s /e ----
             Ran "C:\7-zip\7z.exe" e "C:\Users\vagrant\AppData\Local\Temp\kitchen\cache/javaagent.zip" -o"C:/Users/vagrant/AppData/Local/Temp/d20161221-2696-10pu6l" -uy && for /f %1 in ('dir /ad /b "C:\Users\vagrant\AppData\Local\Temp\d20161221-2696-10pu6l"') do xcopy "C:\Users\vagrant\AppData\Local\Temp\d20161221-2696-10pu6l\%1" "" /s /e returned 1

             Resource Declaration:
             ---------------------
             # In C:/Users/vagrant/AppData/Local/Temp/kitchen/cache/cookbooks/ark/providers/default.rb

             127:   execute "unpack #{new_resource.release_file}" do
             128:     command unpack_command
             129:     cwd new_resource.path
             130:     environment new_resource.environment
             131:     notifies :run, "execute[set owner on #{new_resource.path}]"
             132:     action :nothing
             133:   end
             134:

             Compiled Resource:
             ------------------
             # Declared in C:/Users/vagrant/AppData/Local/Temp/kitchen/cache/cookbooks/ark/providers/default.rb:127:in `block in class_from_file'

             execute("unpack C:\Users\vagrant\AppData\Local\Temp\kitchen\cache/javaagent.zip") do
        action [:nothing]
        retries 0
        retry_delay 2
        default_guard_interpreter :execute
        command "\"C:\\7-zip\\7z.exe\" e \"C:\\Users\\vagrant\\AppData\\Local\\Temp\\kitchen\\cache/javaagent.zip\" -o\"C:/Users/vagrant/AppData/Local/Temp/d20161221-2696-10pu6l\" -uy && for /f %1 in ('dir /ad /b \"C:\\Users\\vagrant\\AppData\\Local\\Temp\\d20161221-2696-10pu6l\"') do xcopy \"C:\\Users\\vagrant\\AppData\\Local\\Temp\\d20161221-2696-10pu6l\\%1\" \"\" /s /e"
        backup 5
        cwd "c:\\appdynamics/javaagent"
        returns 0
        declared_type :execute
        cookbook_name "appdynamics"
             end

             Platform:
             ---------
             x64-mingw32


           ================================================================================
           Error executing action `put` on resource 'ark[javaagent]'
           ================================================================================

           Mixlib::ShellOut::ShellCommandFailed
           ------------------------------------
           execute[unpack C:\Users\vagrant\AppData\Local\Temp\kitchen\cache/javaagent.zip] (C:/Users/vagrant/AppData/Local/Temp/kitchen/cache/cookbooks/ark/providers/default.rb line 127) had an error: Mixlib::ShellOut::ShellCommandFailed: Expected process to exit with [0], but received '1'
           ---- Begin output of "C:\7-zip\7z.exe" e "C:\Users\vagrant\AppData\Local\Temp\kitchen\cache/javaagent.zip" -o"C:/Users/vagrant/AppData/Local/Temp/d20161221-2696-10pu6l" -uy && for /f %1 in ('dir /ad /b "C:\Users\vagrant\AppData\Local\Temp\d20161221-2696-10pu6l"') do xcopy "C:\Users\vagrant\AppData\Local\Temp\d20161221-2696-10pu6l\%1" "" /s /e ----
           STDOUT:
           STDERR: The system cannot find the path specified.
           ---- End output of "C:\7-zip\7z.exe" e "C:\Users\vagrant\AppData\Local\Temp\kitchen\cache/javaagent.zip" -o"C:/Users/vagrant/AppData/Local/Temp/d20161221-2696-10pu6l" -uy && for /f %1 in ('dir /ad /b "C:\Users\vagrant\AppData\Local\Temp\d20161221-2696-10pu6l"') do xcopy "C:\Users\vagrant\AppData\Local\Temp\d20161221-2696-10pu6l\%1" "" /s /e ----
           Ran "C:\7-zip\7z.exe" e "C:\Users\vagrant\AppData\Local\Temp\kitchen\cache/javaagent.zip" -o"C:/Users/vagrant/AppData/Local/Temp/d20161221-2696-10pu6l" -uy && for /f %1 in ('dir /ad /b "C:\Users\vagrant\AppData\Local\Temp\d20161221-2696-10pu6l"') do xcopy "C:\Users\vagrant\AppData\Local\Temp\d20161221-2696-10pu6l\%1" "" /s /e returned 1

           Resource Declaration:
           ---------------------
           # In C:/Users/vagrant/AppData/Local/Temp/kitchen/cache/cookbooks/appdynamics/recipes/java_agent.rb

            23: ark 'javaagent' do
            24:   url "file:///#{Chef::Config[:file_cache_path]}/AppDynamicsJavaAgent.zip"
            25:   path agent['install_dir']
            26:   owner agent['owner']
            27:   action :put
            28: end
            29:

           Compiled Resource:
           ------------------
           # Declared in C:/Users/vagrant/AppData/Local/Temp/kitchen/cache/cookbooks/appdynamics/recipes/java_agent.rb:23:in `from_file'

           ark("javaagent") do
             action [:put]
             updated true
             updated_by_last_action true
             retries 0
             retry_delay 2
             default_guard_interpreter :default
             declared_type :ark
             cookbook_name "appdynamics"
             recipe_name "java_agent"
             url "file:///C:\\Users\\vagrant\\AppData\\Local\\Temp\\kitchen\\cache/AppDynamicsJavaAgent.zip"
             path "c:\\appdynamics/javaagent"
             owner "root"
             extension "zip"
             release_file "C:\\Users\\vagrant\\AppData\\Local\\Temp\\kitchen\\cache/javaagent.zip"
           end

           Platform:
           ---------
           x64-mingw32


       Running handlers:
       [2016-12-21T21:55:39+00:00] ERROR: Running exception handlers
       Running handlers complete
       [2016-12-21T21:55:39+00:00] ERROR: Exception handlers complete
       Chef Client failed. 1 resources updated in 06 seconds
       [2016-12-21T21:55:39+00:00] FATAL: Stacktrace dumped to C:/Users/vagrant/AppData/Local/Temp/kitchen/cache/chef-stacktrace.out
       [2016-12-21T21:55:39+00:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
       [2016-12-21T21:55:39+00:00] FATAL: Mixlib::ShellOut::ShellCommandFailed: ark[javaagent] (appdynamics::java_agent line 23) had an error: Mixlib::ShellOut::ShellCommandFailed: execute[unpack C:\Users\vagrant\AppData\Local\Temp\kitchen\cache/javaagent.zip] (C:/Users/vagrant/AppData/Local/Temp/kitchen/cache/cookbooks/ark/providers/default.rb line 127) had an error: Mixlib::ShellOut::ShellCommandFailed: Expected process to exit with [0], but received '1'
       ---- Begin output of "C:\7-zip\7z.exe" e "C:\Users\vagrant\AppData\Local\Temp\kitchen\cache/javaagent.zip" -o"C:/Users/vagrant/AppData/Local/Temp/d20161221-2696-10pu6l" -uy && for /f %1 in ('dir /ad /b "C:\Users\vagrant\AppData\Local\Temp\d20161221-2696-10pu6l"') do xcopy "C:\Users\vagrant\AppData\Local\Temp\d20161221-2696-10pu6l\%1" "" /s /e ----
       STDOUT:
       STDERR: The system cannot find the path specified.
       ---- End output of "C:\7-zip\7z.exe" e "C:\Users\vagrant\AppData\Local\Temp\kitchen\cache/javaagent.zip" -o"C:/Users/vagrant/AppData/Local/Temp/d20161221-2696-10pu6l" -uy && for /f %1 in ('dir /ad /b "C:\Users\vagrant\AppData\Local\Temp\d20161221-2696-10pu6l"') do xcopy "C:\Users\vagrant\AppData\Local\Temp\d20161221-2696-10pu6l\%1" "" /s /e ----
       Ran "C:\7-zip\7z.exe" e "C:\Users\vagrant\AppData\Local\Temp\kitchen\cache/javaagent.zip" -o"C:/Users/vagrant/AppData/Local/Temp/d20161221-2696-10pu6l" -uy && for /f %1 in ('dir /ad /b "C:\Users\vagrant\AppData\Local\Temp\d20161221-2696-10pu6l"') do xcopy "C:\Users\vagrant\AppData\Local\Temp\d20161221-2696-10pu6l\%1" "" /s /e returned 1
```

**Comment by spuder** [comment]
In 1.1.0 of the ark cookbook, 7zip is hard coded to c:\7-zip\7-zip
https://github.com/chef-cookbooks/ark/blob/v1.1.0/attributes/default.rb#L25

In 1.2.0 of the ark cookbook, it fetches the path from the registry

https://github.com/chef-cookbooks/ark/blob/v1.2.0/attributes/default.rb#L25

I remember pinning on version 1.1.0 because something broke in 1.2.0. Since we install 7zip from chocolatey, this won't work.

**Comment by dkoepke** [comment]
I believe this was fixed in v0.3.2.
